jmeter/src/main/java/org/eclipse/hono/jmeter/client/HonoCommander.java

```diff
@@ -0,0 +1,259 @@
+    }
+
+    private void handleMessage(final String msgType, final Message msg) {
+        final Data body = (Data) msg.getBody();
+        LOG.debug("Type: [{}] and Message: [{}]", msgType, body != null ? body.getValue().toString() : "");
+    }
+
+    private void handleCommandReadinessNotification(final TimeUntilDisconnectNotification notification) {
+        if (notification.getMillisecondsUntilExpiry() == 0) {
+            LOG.trace("Device [{}:{}] not ready to receive commands.", notification.getTenantId(),
+                    notification.getDeviceId());
+            if ("sampler".equals(sampler.getTriggerType())) {
+                devicesReadyToReceiveCommands.remove(notification.getDeviceId());
+            }
+        } else {
+            if ("device".equals(sampler.getTriggerType())) {
+                LOG.debug("Sending command to device [{}:{}]", notification.getTenantId(), notification.getDeviceId());
+                setSampleStartIfNotSetYet(System.currentTimeMillis());
+                sendCommandAndReceiveResponse(notification.getTenantId(), notification.getDeviceId());
+            } else {
+                LOG.debug("A device [{}] got registered with trigger type [{}]", notification.getDeviceId(),
+                        sampler.getTriggerType());
+                devicesReadyToReceiveCommands.add(notification.getDeviceId());
+            }
+        }
+    }
+
+    private void sendCommandAndReceiveResponse(final String tenantId, final String deviceId) {
+        client.getOrCreateCommandClient(tenantId, deviceId)
+                .map(this::setCommandTimeOut)
+                .compose(commandClient -> commandClient
+                        .sendCommand(sampler.getCommand(), Buffer.buffer(sampler.getCommandPayload()))
+                        .map(commandResponse -> {
+                            final String commandResponseText = Optional.ofNullable(commandResponse.getPayload())
+                                    .orElse(Buffer.buffer()).toString();
+                            LOG.debug("Command response from device [{}:{}] received [{}]", tenantId,
+                                    deviceId, commandResponseText);
+                            synchronized (lock) {
+                                successCount.addAndGet(1);
+                                bytesReceived.addAndGet(sampler.getCommandPayload().getBytes().length);
+                                bytesSent.addAndGet(commandResponseText.getBytes().length);
+                            }
+                            return commandResponse;
+                        })
+                        .map(x -> closeCommandClient(commandClient, tenantId, deviceId))
+                        .recover(error -> {
+                            if (sampler.getTriggerType().equals("device")
+                                    || devicesReadyToReceiveCommands.contains(deviceId)) {
+                                errorCount.addAndGet(1);
```

Comment:
Not synchronized.

Reply:
synchronized (lock) {
	successCount.addAndGet(1);
	bytesReceived.addAndGet(sampler.getCommandPayload().getBytes().length);
	bytesSent.addAndGet(commandResponseText.getBytes().length);
}
In the above part, I used synchronized block to let one thread to work on all the three lines at a time. In case of errorCount.addAndGet(1) there is only one line and hence there is no synchronised is needed and only vertx threads work on these parts.

Reply:
See comment above - FMPOV it need to be synchronized with the code in the sample() method.

Reply:
pls see comment above.